Content.Server/Abilities/Psionics/PsionicAbilitiesSystem.cs

```diff
@@ -62,66 +77,185 @@ public void AddPsionics(EntityUid uid, bool warn = true)
-                return;
+                if (!_prototypeManager.TryIndex<PsionicPowerPrototype>(proto.ID, out var power))
+                    continue;
+
+                if (power.Components is not null)
+                    foreach (var comp in power.Components)
+                        if (EntityManager.TryGetComponent(uid, _componentFactory.GetComponent(comp).GetType(), out var powerComp)
+                            && powerComp is not null)
+                            RemComp(uid, powerComp);
+
+                if (psionic.Actions is not null)
+                    foreach (var action in psionic.Actions)
+                        _actionsSystem.RemoveAction(uid, action.Entity);
+
+                // If we're mindbreaking, we can skip the last two enumerations since the PsionicComponent is getting 1984'd.
+                if (mindbreak)
+                    continue;
+
+                if (power.AmplificationModifier != 0)
+                    psionic.AmplificationSources.Remove(power.Name);
+
+                if (power.DampeningModifier != 0)
+                    psionic.DampeningSources.Remove(power.Name);
             }
 
-            foreach (var compName in pool.Weights.Keys)
+            if (mindbreak)
             {
-                // component moment
-                var comp = _componentFactory.GetComponent(compName);
-                if (EntityManager.TryGetComponent(uid, comp.GetType(), out var psionicPower))
-                    RemComp(uid, psionicPower);
+                RemComp<PsionicComponent>(uid);
+                EnsureComp<MindbrokenComponent>(uid);
+                _statusEffectsSystem.TryAddStatusEffect(uid, "Stutter", TimeSpan.FromMinutes(5 * psionic.CurrentAmplification * psionic.CurrentDampening), false, "StutteringAccent");
             }
-            if (psionic.PsionicAbility != null
-                && _actionsSystem.TryGetActionData(psionic.PsionicAbility, out var psiAbility)
-                && psiAbility is not null)
-                _actionsSystem.RemoveAction(uid, psionic.PsionicAbility);
+        }
+
+        public void RemovePsionicPower(EntityUid uid, PsionicPowerPrototype proto, bool removedByComponent = false)
+        {
+            if (!TryComp<PsionicComponent>(uid, out var psionic)
+                || !psionic.Removable
+                || !_prototypeManager.TryIndex(proto.ID, out var _))
+                return;
+
+            if (removedByComponent
+                && proto.Components is not null)
+                foreach (var comp in proto.Components)
+                    if (EntityManager.TryGetComponent(uid, _componentFactory.GetComponent(comp).GetType(), out var powerComp)
+                        && powerComp is not null)
+                        RemComp(uid, powerComp);
+
+            if (psionic.Actions is not null)
+                foreach (var action in psionic.Actions)
+                    _actionsSystem.RemoveAction(uid, action.Entity);
```

Comment:
these loops/ifs could benefit from braces for readability

Reply:
These blocks also seem similar, if not identical, to those at line 215 and above. Is it possible to extract them to a local function to avoid redundant code?